File: backend/app/api/operator.py
"""
GreenCharge — Operator API Router (P4)

Endpoints (api-contract.md SS4, SS5):
    GET /api/stations
    GET /api/chargers
    GET /api/network/status
    GET /api/network/impact

This file only wires HTTP <-> services. No business logic here.
Optimization run/apply are P2-owned (POST /api/optimization/run,
POST /api/optimization/apply) — this router does not duplicate them;
the frontend calls P2's optimization client directly.

Registration (main.py, append-only one line):
    from app.api.operator import router as operator_router
    app.include_router(operator_router, prefix="/api", tags=["operator"])
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.database import get_db

from app.services.operator.stations import get_stations
from app.services.operator.chargers import get_chargers
from app.services.operator.network import (
    get_network_status,
    get_network_impact,
    NoActiveScheduleError,
)

from app.schemas.operator import (
    StationsResponse,
    ChargersResponse,
    NetworkStatusResponse,
    NetworkImpactResponse,
)

router = APIRouter()


@router.get("/stations", response_model=StationsResponse)
def read_stations(db: Session = Depends(get_db)):
    return get_stations(db)


@router.get("/chargers", response_model=ChargersResponse)
def read_chargers(
    station_id: str | None = Query(default=None), db: Session = Depends(get_db)
):
    return get_chargers(db, station_id=station_id)


@router.get("/network/status", response_model=NetworkStatusResponse)
def read_network_status(db: Session = Depends(get_db)):
    return get_network_status(db)


@router.get("/network/impact", response_model=NetworkImpactResponse)
def read_network_impact(db: Session = Depends(get_db)):
    try:
        return get_network_impact(db)
    except NoActiveScheduleError as exc:
        raise HTTPException(
            status_code=409,
            detail={"error": {"code": "NO_ACTIVE_SCHEDULE", "message": str(exc)}},
        )

File: frontend/src/services/operator.ts
/**
 * GreenCharge — Operator API Client (P4)
 *
 * Thin typed wrapper over /api/stations, /api/chargers, /api/network/*.
 * Uses shared types from frontend/src/types/api.ts only (rules.md SS12).
 *
 * Does NOT wrap /api/optimization/run or /api/optimization/apply — those
 * are P2-owned (frontend/src/services/optimization.ts). The operator
 * pages import that client directly rather than re-wrapping it here,
 * so there is exactly one client per shared endpoint.
 */

import type {
  StationsResponse,
  ChargersResponse,
  NetworkStatusResponse,
  NetworkImpactResponse,
  ErrorResponse,
} from "../types/api";

async function handle<T>(res: Response): Promise<T> {
  if (!res.ok) {
    let message = `Request failed (${res.status})`;
    try {
      const body = (await res.json()) as ErrorResponse & {
        detail?: string | { error?: ErrorResponse["error"] };
      };
      if (typeof body.detail === "string") {
        message = body.detail;
      } else if (body.detail?.error?.message) {
        message = body.detail.error.message;
      } else if (body.error?.message) {
        message = body.error.message;
      }
    } catch {
      // Keep the generic HTTP error.
    }
    throw new Error(message);
  }
  return res.json() as Promise<T>;
}

export async function getStations(): Promise<StationsResponse> {
  const res = await fetch("/api/stations");
  return handle<StationsResponse>(res);
}

export async function getChargers(stationId?: string): Promise<ChargersResponse> {
  const url = stationId
    ? `/api/chargers?station_id=${encodeURIComponent(stationId)}`
    : "/api/chargers";
  const res = await fetch(url);
  return handle<ChargersResponse>(res);
}

export async function getNetworkStatus(): Promise<NetworkStatusResponse> {
  const res = await fetch("/api/network/status");
  return handle<NetworkStatusResponse>(res);
}

export async function getNetworkImpact(): Promise<NetworkImpactResponse> {
  const res = await fetch("/api/network/impact");
  return handle<NetworkImpactResponse>(res);
}
